### ISSUE TITLE
Include an example of how to use a controller-decoupled ahoy

### ISSUE THREAD

**Issue by Startouf** [ankane/ahoy#338]
It would be nice to add an example showing how to use ahoy without the context of a controller/request.

For instance, while doing maintenance in a rails console, during tests, etc. we need to be able to call `ahoy.track` from some ruby context other than a controller

I originally though `ahoy = Ahoy::Tracker.new(controller: nil)` would be enough but then, given my following initializer

```
class Ahoy::Store < Ahoy::Stores::MongoidStore
  def track_event(name, properties, options)
    super do |event|
      event.ip = request.ip
      event.referrer = request.referrer
    end
  end
end
```

THe code crashes with 
> undefined method `ip' for nil:NilClass (NoMethodError)

So some precautions must be taken... ie `request&.ip` (or `request && request.ip` on old rubies) and the readme should be updated

**Comment by ankane** [comment]
Hey @Startouf, Ahoy isn't meant to be used outside the context of a controller. Can you give more details on your use case?

**Comment by Startouf** [comment]
@ankane I have several use cases

1. Asynchronous side-effect processing. I have accumulate business logic on my services providing synchronous response and response times are getting quite high, so I'm moving most side-effects ( including Ahoy tracking) to jobs 

2. Maintenance. Often I need to connect to a rails console in production and run similar business logic from the same services that are used by my controllers. Usually, my controllers are passed to my services and I delegate ahoy tracking to the controller, but when doing such console-ops I have no controller to pass, so I need to fallback to some request/controller-less ahoy implementation

In a nutshell, this is what I have

```
class FoosController < APIController
  def create
    service = MyService.new(controller: self)
    service.call
    service.success? render_ok_response else :render_bad_response
  end 
end

class MyService
  def initialize(controller: nil)
    @controller = controller
  end
 
  def call
    ... (business logic)
    track_success
  end

  private

  def track_success
    ahoy.track(...)
    intercom.track(...)
  end

  def ahoy
    controller.present? ? controller.ahoy : controller_decoupled_ahoy
  end
end
 
```

**Comment by ankane** [comment]
Thanks for sharing. For case 1, I'd recommend keeping the tracking in the foreground, even if you want to move the storage and/or additional processing to the background. It should be pretty lightweight.

For case 2, while not officially supported, you can try passing a `nil` controller. The reason it's crashing is your custom tracking code expects a `request` object (not Ahoy itself).

**Comment by danielfriis** [comment]
I'm facing the same use case as 1) - I want to move Ahoy tracking into a background job

**Comment by Startouf** [comment]
Hello @ankane I just updated to ahoy v2 (cannot upgrade directly to v3 as we are still on Ruby 2.3, requiring Ruby >=2.4 is a bit harsh ^^", do you really need it on your v3 ?)

I'm running into new issues when using trackers
> undefined method `cookie_jar' for nil:NilClass

With `::Ahoy::Tracker.new(controller: nil)`

I would suggest modifying [this line](https://github.com/ankane/ahoy/blob/1fbdca705263122222fd72c3e376df816652ee7b/lib/ahoy/tracker.rb#L179) to make it more resilient / compatible with no controller/request

Furthermore, it would be nice to add an attr_reader on the tracker class for the `@user` variable.

Since I need to instanciate an Ahoy tracker from a background job, in order to attach the user to the event (example: I have a "reply by email" service, so there is indeed a user action but no controller involved)

Right now I need to `ahoy.instance_variable_set(:@user, user)` to set manually the user in my service tracker.

**Comment by Startouf** [comment]
Actually a lot more code than just the cookie_jar is broken with newer versions of ahoy, since there are so many calls to `request` assuming it is not nil...
